index.js

```diff
@@ -1,389 +1,156 @@
+var monkeypatch = require("./utils/monkeypatch");
+var getESLintModules = require("./utils/get-eslint-modules");
+
+/**
+ * Get the ESLint-related modules
+ */
+var modules = getESLintModules();
+var Traverser = modules.Traverser;
+var ScopeManager = modules.ScopeManager;
+
+/**
+ * Configure extendedVisitorKeys, based on the default keys from ESLint and the
+ * VISITOR_KEYS from babel-types
+ */
+var DEFAULT_VISITOR_KEYS = Traverser.DEFAULT_VISITOR_KEYS;
+var extendedVisitorKeys = Object.assign({}, DEFAULT_VISITOR_KEYS, VISITOR_KEYS);
+extendedVisitorKeys.MethodDefinition.push("decorators");
+extendedVisitorKeys.Property.push("decorators");
+
+/**
+ * Globals
+ */
 var hasPatched = false;
 var eslintOptions = {};
+var EnhancedReferencer = makeEnhancedReferencer(modules, extendedVisitorKeys);
+
+/**
+ * Ensure that reasonable default parser options are set.
+ *
+ * @param {Object} options Parser options
+ * @returns {Object} parserOptions with defaults set if appropriate
+ */
+function ensureDefaultParserOptions(options) {
+  var parserOptions = (options = options || {});
+  parserOptions.ecmaVersion = options.ecmaVersion || 6;
+  parserOptions.sourceType = options.sourceType || "module";
+  parserOptions.allowImportExportEverywhere =
+    options.allowImportExportEverywhere || false;
+  return parserOptions;
+}
 
-function getModules() {
-  try {
-    // avoid importing a local copy of eslint, try to find a peer dependency
-    var eslintLoc = Module._resolveFilename("eslint", module.parent);
-  } catch (err) {
-    try {
-      // avoids breaking in jest where module.parent is undefined
-      eslintLoc = require.resolve("eslint");
-    } catch (err) {
-      throw new ReferenceError("couldn't resolve eslint");
-    }
-  }
-
-  // get modules relative to what eslint will load
-  var eslintMod = new Module(eslintLoc);
-  eslintMod.filename = eslintLoc;
-  eslintMod.paths = Module._nodeModulePaths(path.dirname(eslintLoc));
-
-  try {
-    var escope = eslintMod.require("eslint-scope");
-    var Definition = eslintMod.require("eslint-scope/lib/definition")
-      .Definition;
-    var referencer = eslintMod.require("eslint-scope/lib/referencer");
-  } catch (err) {
-    escope = eslintMod.require("escope");
-    Definition = eslintMod.require("escope/lib/definition").Definition;
-    referencer = eslintMod.require("escope/lib/referencer");
+/**
+ * Use the given parser options to inform the global ESLint configuration options.
+ *
+ * @param {Object} parserOptions Parser options
+ * @returns {void}
+ */
+function setESLintOptions(parserOptions) {
+  eslintOptions.ecmaVersion = parserOptions.ecmaVersion;
+  eslintOptions.sourceType = parserOptions.sourceType;
+  eslintOptions.allowImportExportEverywhere =
+    parserOptions.allowImportExportEverywhere;
+  if (parserOptions.sourceType === "module") {
+    eslintOptions.globalReturn = false;
+  } else {
+    delete eslintOptions.globalReturn;
```

Comment:
`babel-eslint` seems to enable `allowReturnOutsideFunction` option always, so I guess `eslintOptions.ecmaFeatures.globalReturn` should be `true`. Or `allowReturnOutsideFunction` should follow to `eslintOptions.ecmaFeatures.globalReturn`.